core/src/main/java/org/apache/iceberg/encryption/KeyMetadataEncoder.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.encryption;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import org.apache.avro.io.BinaryEncoder;
+import org.apache.avro.io.DatumWriter;
+import org.apache.avro.io.EncoderFactory;
+import org.apache.avro.message.MessageEncoder;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.avro.AvroSchemaUtil;
+import org.apache.iceberg.data.avro.DataWriter;
+
+public class KeyMetadataEncoder<D> implements MessageEncoder<D> {
+  private final byte schemaVersion;
+  private final DatumWriter<D> writer;
+  private final BufferOutputStream bufferOutputStream;
+  private BinaryEncoder binaryEncoder;
+
+  /**
+   * Creates a new {@link MessageEncoder} that will deconstruct datum instances described by the
+   * {@link Schema schema}.
+   *
+   * <p>Buffers returned by {@code encode} are copied and will not be modified by future calls to
+   * {@code encode}.
+   *
+   * @param schema the {@link Schema} for datum instances
+   */
+  public KeyMetadataEncoder(byte version, Schema schema) {
```

Comment:
Like the decoder, I think this should have a static set of schemas, probably stored in `KeyMetadata`, rather than passing one in. It's okay to have an argument for the schema version to produce (1) but even that isn't really necessary right now.